src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SDSEarlyReleaseDefaultingRulesService.kt

```diff
@@ -29,144 +29,198 @@ class SDSEarlyReleaseDefaultingRulesService(
-      val controllingDate = getControllingDate(dates)
-
-      dates[ReleaseDateType.HDCED]?.let { hdcedDate ->
-        val prrdDate = dates[ReleaseDateType.PRRD] ?: otherDates[ReleaseDateType.PRRD]
-        prrdDate?.takeIf { it.isAfter(hdcedDate) }?.let {
-          setHDCEDDates(it, dates)
-          standardReleaseResult.breakdownByReleaseDateType[ReleaseDateType.HDCED]?.apply {
-            breakdownByReleaseDateType[ReleaseDateType.HDCED] = copy(
-              rules = rules + CalculationRule.HDCED_ADJUSTED_TO_CONCURRENT_PRRD,
-            )
-          }
-        }
-      }
+      val (ardOrCrdType, ardOrCrdDate) = getArdOrCrd(dates)
 
-      if (dates.containsKey(ReleaseDateType.PRRD)) {
-        if (controllingDate.second?.isAfter(dates[ReleaseDateType.PRRD]) == true) {
-          dates.remove(ReleaseDateType.PRRD)
-        } else {
-          dates.remove(controllingDate.first)
+      adjustHdcedForPrrd(dates, otherDates, breakdownByReleaseDateType, standardReleaseResult)
+      adjustPrrdForArdOrCrd(dates, ardOrCrdType, ardOrCrdDate)
+    }
+  }
 
-          // PRRD is later than any other release date, therefore can not have a HDCED
-          // See BookingExtractionService#extractMultiple
-          removeHDCEDDates(dates)
-        }
-      }
+  private fun adjustHdcedForPrrd(
+    dates: MutableMap<ReleaseDateType, LocalDate>,
+    otherDates: MutableMap<ReleaseDateType, LocalDate>,
+    breakdownByReleaseDateType: MutableMap<ReleaseDateType, ReleaseDateCalculationBreakdown>,
+    standardReleaseResult: CalculationResult,
+  ) {
+    val hdcedDate = dates[ReleaseDateType.HDCED] ?: return
+    val prrdDate = dates[ReleaseDateType.PRRD] ?: otherDates[ReleaseDateType.PRRD] ?: return
+
+    if (prrdDate.isAfter(hdcedDate)) {
+      updateHdcedDateAndBreakdown(dates, prrdDate, breakdownByReleaseDateType, standardReleaseResult)
     }
   }
 
-  private fun setHDCEDDates(date: LocalDate, dates: MutableMap<ReleaseDateType, LocalDate>) {
-    dates[ReleaseDateType.HDCED] = date
+  private fun updateHdcedDateAndBreakdown(
+    dates: MutableMap<ReleaseDateType, LocalDate>,
+    newDate: LocalDate,
+    breakdownByReleaseDateType: MutableMap<ReleaseDateType, ReleaseDateCalculationBreakdown>,
+    standardReleaseResult: CalculationResult,
+  ) {
+    dates[ReleaseDateType.HDCED] = newDate
+    standardReleaseResult.breakdownByReleaseDateType[ReleaseDateType.HDCED]?.let { standardBreakdown ->
+      breakdownByReleaseDateType[ReleaseDateType.HDCED] = standardBreakdown.copy(
+        rules = standardBreakdown.rules + CalculationRule.HDCED_ADJUSTED_TO_CONCURRENT_PRRD,
+      )
+    }
   }
 
-  private fun removeHDCEDDates(dates: MutableMap<ReleaseDateType, LocalDate>) {
-    dates.remove(ReleaseDateType.HDCED)
+  private fun adjustPrrdForArdOrCrd(
+    dates: MutableMap<ReleaseDateType, LocalDate>,
+    ardOrCrdType: ReleaseDateType,
+    ardOrCrdDate: LocalDate?,
+  ) {
+    if (dates.containsKey(ReleaseDateType.PRRD)) {
```

Comment:
Should this condition include ReleaseDateType.HDCED? I assume this is only applicable if PRRD is present.

Reply:
good question, going by the assumption that the old code was correct then yes.. I havent changed any functionality just refactored. but agree it does sound a bit odd at this level, but there are some weird rules in CRD - so very pkausible. Ive pasted the old code below that did this bit.

```

      if (dates.containsKey(ReleaseDateType.PRRD)) {
        if (controllingDate.second?.isAfter(dates[ReleaseDateType.PRRD]) == true) {
          dates.remove(ReleaseDateType.PRRD)
        } else {
          dates.remove(controllingDate.first)

          // PRRD is later than any other release date, therefore can not have a HDCED
          // See BookingExtractionService#extractMultiple
          removeHDCEDDates(dates)
        }
      }
```